jmdaemon/jmdaemon/orderbookwatch.py

```diff
@@ -134,17 +139,74 @@ def on_order_cancel(self, counterparty, oid):
         finally:
             self.dblock.release()
 
+    def on_fidelity_bond_seen(self, nick, bond_type, fidelity_bond_b64data):
+        try:
+            fidelity_bond_data = base64.b64decode(fidelity_bond_b64data)
+        except ValueError as e:
+            log.debug("error parsing base64-encoding: " + repr(e) + ", ignoring")
+            return
+
+        #nick_sig + cert_sig + cert_pubkey + cert_expiry + utxo_pubkey + txid + vout + timelock
+        #72       + 72       + 33          + 2           + 33          + 32   + 4    + 4 = 252bytes
+        data_blob_length = 252
+
+        if len(fidelity_bond_data) != data_blob_length:
+            log.debug("fidelity bond data from " + nick + " wrong length: "
+                + str(len(fidelity_bond_data)) + ", ignoring")
+            return
+        try:
+            (nick_signature, certificate_signature, certificate_pubkey,
+                certificate_expiry, utxo_pubkey, txid, vout,
+                locktime) = struct.unpack("<72s72s33sH33s32sII", fidelity_bond_data)
+        except struct.error as e:
+            log.debug("unable to unpack fidelity bond data: " + str(len(fidelity_bond_data))
+                + " e=" + repr(e) + ", ignoring")
+            return
+
+        try:
+            #remove padding
+            #the DER signature format always has a initial \x30 byte as the header
```

Comment:
Re: "remove padding":

`jmbitcoin.ecdsa_verify` certainly does expect DER encoding, including the leading 0x30 byte.

You're not removing padding. When you do `x[x.index(b"\x30"]:]` you're just keeping `x`, not removing the first byte.

Reply:
This DER part is weird. The signature is generated by `ecdsa_sign` and verified by `ecdsa_verify`. I don't think that should need any hacks in between.

Reply:
The purpose of the padding is to make the signatures always be the same length (72 bytes). `ecdsa_sign` will produce signatures of different lengths, occasionally 71, 70 or 69.

Reply:
Of course. I didn't realise that was removing the `0xff` bytes.

Reply:
Oh I get it now, sorry. I had actually read the part where you padded to 72 bytes but didn't put two and two together :)